drivers/spi/spi_sy1xx.c

```diff
@@ -0,0 +1,506 @@
+
+	if (ret < 0) {
+		LOG_ERR("Failed to prepare transfer");
+		return -EINVAL;
+	}
+
+	if (ret == 0) {
+		LOG_WRN("Nothing to do");
+		return 0;
+	}
+
+	/* expected data config and spi transfer type */
+	buf[count++] = SPI_CMD_SEND_DATA3(request->data_len * 8);
+	buf[count++] = SPI_CMD_SEND_DATA2(request->data_len * 8);
+	buf[count++] = SPI_CMD_SEND_DATA1();
+	buf[count++] = SPI_CMD_SEND_DATA0(config->quad_spi);
+
+	/* data; we need to fill multiple of 32bit size */
+	uint32_t word_count = ROUND_UP_DIV_4(request->data_len);
+
+	for (uint32_t i = 0; i < request->data_len; i++) {
+		buf[count + i] = request->data[i];
+	}
+	count += (word_count * 4);
+
+	/* end of transmission */
+	buf[count++] = SPI_CMD_EOT3(0);
+	buf[count++] = SPI_CMD_EOT2();
+	buf[count++] = SPI_CMD_EOT1();
+	buf[count++] = SPI_CMD_EOT0();
+
+	SY1XX_UDMA_START_TX(config->base, (uint32_t)buf, count, 0);
+
+	SY1XX_UDMA_WAIT_FOR_FINISHED_TX(config->base);
+
+	int32_t xfer_count = request->data_len - SY1XX_UDMA_GET_REMAINING_RX(config->base);
+
+	if ((xfer_count > 0) && (xfer_count <= request->data_len)) {
+		return xfer_count;
+	}
+
+	return 0;
+}
+
+/**
+ * @return:
+ *  - < 0: Error
+ *  - = 0: No bytes received
+ *  - > 0: Number of bytes received
+ */
+static int32_t sy1xx_spi_read(const struct device *dev, struct sy1xx_spi_xfer *request)
+{
+
+	struct sy1xx_spi_config *config = (struct sy1xx_spi_config *)dev->config;
+	struct sy1xx_spi_data *data = (struct sy1xx_spi_data *)dev->data;
+
+	uint8_t *buf = &data->write[0];
+
+	uint32_t count = 0;
+
+	/* prepare start of transfer and address */
+	int32_t ret = sy1xx_spi_prepare_header(dev, buf, &count, request);
+
+	if (ret < 0) {
+		LOG_ERR("Failed to prepare transfer");
+		return -EINVAL;
+	}
+
+	if (ret == 0) {
+		LOG_WRN("Nothing to do");
+		return 0;
+	}
+
+	/* expected data config */
+	buf[count++] = SPI_CMD_READ_DATA3(request->data_len * 8);
+	buf[count++] = SPI_CMD_READ_DATA2(request->data_len * 8);
+	buf[count++] = SPI_CMD_READ_DATA1();
+	buf[count++] = SPI_CMD_READ_DATA0(0, 0);
+
+	/* end of transmission */
+	buf[count++] = SPI_CMD_EOT3(0);
+	buf[count++] = SPI_CMD_EOT2();
+	buf[count++] = SPI_CMD_EOT1();
+	buf[count++] = SPI_CMD_EOT0();
+
+	SY1XX_UDMA_START_RX(config->base, (uint32_t)data->read, request->data_len,
+			    SY1XX_UDMA_RX_DATA_ADDR_INC_SIZE_32);
+	SY1XX_UDMA_START_TX(config->base, (uint32_t)buf, count, 0);
+
+	SY1XX_UDMA_WAIT_FOR_FINISHED_TX(config->base);
+	SY1XX_UDMA_WAIT_FOR_FINISHED_RX(config->base);
+
+	int32_t xfer_count = request->data_len - SY1XX_UDMA_GET_REMAINING_RX(config->base);
+
+	if ((xfer_count > 0) && (xfer_count <= request->data_len)) {
+		memcpy(request->data, &data->read[0], xfer_count);
+		return xfer_count;
+	}
+
+	return 0;
+}
+
+/**
+ * the problem with transfer function is, that it comes in
+ * multiple chunks. where we actually want to use DMA to transfer all address + data
+ * at once.
+ *
+ * transfers could be,
```

Comment:
What are you trying to implement? that does not look like a generic SPI controller driver to me. 

address+data format of transaction is something not generic, that one can find in spi flash transactions (or eeproms etc...)

Reply:
lets say there is a read of some bytes from defined address. then this would be 2 transfers. however this spi controller can do this in one transfer, even more it does it in DMA.